include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -673,22 +819,51 @@ struct __gen_set_op_from_known_balanced_path
         _SizeType __i_elem = __id * __diagonal_spacing;
         if (__i_elem >= __rng1.size() + __rng2.size())
             return std::make_tuple(std::uint32_t{0}, std::uint16_t{0});
-        _SizeType __star_offset = oneapi::dpl::__internal::__dpl_signbit(__rng1_temp_diag[__id]) ? 1 : 0;
-        auto __rng1_temp_diag_abs = std::abs(__rng1_temp_diag[__id]);
-        auto __rng2_temp_diag = __i_elem - __rng1_temp_diag_abs + __star_offset;
+        auto [__rng1_idx, __rng2_idx, __star_offset] =
+            oneapi::dpl::__par_backend_hetero::__decode_balanced_path_temp_data(__rng1_temp_diag, __id,
+                                                                                __diagonal_spacing);
 
         _SizeType __eles_to_process =
-            std::min(_SizeType{__diagonal_spacing} - __star_offset, __rng1.size() + __rng2.size() - (__i_elem - 1));
+            std::min(_SizeType{__diagonal_spacing - __star_offset}, __rng1.size() + __rng2.size() - (__i_elem - 1));
 
-        std::uint16_t __count = __set_op_count(__rng1, __rng2, __rng1_temp_diag_abs, __rng2_temp_diag,
-                                               __eles_to_process, __output_data, __comp);
+        std::uint16_t __count =
+            __set_op_count(__rng1, __rng2, __rng1_idx, __rng2_idx, __eles_to_process, __output_data, __comp);
         return std::make_tuple(std::uint32_t{__count}, __count);
     }
     _SetOpCount __set_op_count;
     std::uint16_t __diagonal_spacing;
     _Compare __comp;
 };
 
+// kernel for balanced path to partition the input into tiles by calculating balanced path on diagonals of tile bounds
+template <typename _GenInput, typename _KernelName>
+struct __partition_set_balanced_path_submitter;
+template <typename _GenInput, typename... _KernelName>
+struct __partition_set_balanced_path_submitter<_GenInput, __internal::__optional_kernel_name<_KernelName...>>
+{
+    template <typename _InInOutRng>
+    sycl::event
+    operator()(sycl::queue& __q, _InInOutRng&& __in_in_out_rng, std::size_t __num_diagonals) const
+    {
+        const std::size_t __tile_size = __gen_input.__get_bounds.__tile_size;
+        const std::size_t __n =
+            oneapi::dpl::__internal::__dpl_ceiling_div(__num_diagonals + __tile_size - 1, __tile_size);
+        return __q.submit([&__in_in_out_rng, this, __n, __num_diagonals](sycl::handler& __cgh) {
```

Comment:
Since we are capturing `this` and we have a device copyable specialization for `__gen_set_balanced_path` which is a member of this class, I think we may need a device copyable specialization for `__partition_set_balanced_path_submitter` as well.

Reply:
good catch.  fixed.